From 04173e7f3ddaec11ba046d058c785107554d0921 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:04:57 -0400
Subject: [fs] btrfs: only unlock block in verify_parent_transid if we locked it

Message-id: <1407179109-13714-411-git-send-email-sandeen@redhat.com>
Patchwork-id: 87287
O-Subject: [RHEL7.1 PATCH 410/422] btrfs: only unlock block in verify_parent_transid if we locked it
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Josef Bacik <jbacik@fb.com>

commit 472b909ff6f4884d235ef7b9d3847fad5efafbff
Author: Josef Bacik <jbacik@fb.com>
Date:   Wed Jun 25 13:45:41 2014 -0700

    btrfs: only unlock block in verify_parent_transid if we locked it

    This is a regression from my patch a26e8c9f75b0bfd8cccc9e8f110737b136eb5994, we
    need to only unlock the block if we were the one who locked it.  Otherwise this
    will trip BUG_ON()'s in locking.c  Thanks,

    cc: stable@vger.kernel.org
    Signed-off-by: Josef Bacik <jbacik@fb.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/disk-io.c b/fs/btrfs/disk-io.c
index 9bc8ce6..dedecf8 100644
--- a/fs/btrfs/disk-io.c
+++ b/fs/btrfs/disk-io.c
@@ -369,7 +369,8 @@ static int verify_parent_transid(struct extent_io_tree *io_tree,
 out:
  unlock_extent_cached(io_tree, eb->start, eb->start + eb->len - 1,
         &cached_state, GFP_NOFS);
- btrfs_tree_read_unlock_blocking(eb);
+ if (need_lock)
+  btrfs_tree_read_unlock_blocking(eb);
  return ret;
 }
 
-- 
1.7.1